src/transformStateWithClones.js

```diff
@@ -6,8 +6,36 @@
  *
  * @return {Object[]}
  */
+
 function transformStateWithClones(state, actions) {
-  // write code here
+  const actionsOfState = [];
+  const stateClone = { ...state };
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        Object.assign(stateClone, action.extraData);
+        break;
+
+      case 'removeProperties':
+        action.keysToRemove.forEach(element => delete stateClone[element]);
+        break;
+
+      case 'clear':
+        for (const key in stateClone) {
+          delete stateClone[key];
+        }
+        break;
+
+      default:
+        Error('sth gone wrong :(');
+        break;
+    }
+
+    actionsOfState.push(Object.assign({}, stateClone));
```

Comment:
`actionOfState.push({ ...stateClone });` also should work so I guess it's up to you with one prefer